doc/developer_guide/HNS_error_bestiary.md

```diff
@@ -1,15 +1,120 @@
 # HNS errors bestiary
 
-HNS error messages ofter aren't descriptive enough. 
-Here are error messages the team encountered in different contexts, and their meaning.
+HNS error messages ofter aren't descriptive enough, and the cleanup is quite hard.
+The following document describes common cleanup techniques and error interpretations.
+
+## HNS Decontamination procedures
+
+Before working with HNS, one must realize the risks.
+There are many ways in which HNS can stop working and affect other parts of the system.
+For a developer, it is important to be able to recognize those, which can be recovered 
+from. Below are some techniques which may help.
+
+Always try cleaning using the lowest level procedure. The higher the level, the more 
+potential damage the cleanup can cause.
+
+### Level 1 decontamination procedure
+
+1. First, stop the docker service.
+```
+Stop-Service docker
+```
+2. Remove all virtual switches, container networks and NAT
+```
+Get-ContainerNetwork | Remove-ContainerNetwork
+Get-VMSwitch | Remove-VMSwitch
+Get-NetNAT | Remove-NetNAT
+```
+
+3. Restart HNS and docker services.
+```
+Restart-Service hns
+Restart-Service docker
+```
+
+### Level 2 decontamination procedure
+
+
+
+1. Remove all container networks. Always do this because some vmswitches may not be removed properly if you just execute the next command.
+```
+Get-ContainerNetwork | Remove-ContainerNetwork
+```
+
+2. Manually remove `HNS.data` file. `net` command sometimes works better at restarting HNS than PowerShell's `Restart-Service`.
+
+> Warning: this is not recommended by Microsoft, but they do it in their official cleanup script, so I guess it's not that bad.
+
+```
+net stop hns; 
+del C:\programdata\Microsoft\Windows\HNS\HNS.data; 
+net start hns;
+```
+
+### Level 3 decontamination procedure
+
+Use official Microsoft script to cleanup. It will also cleanup some registry entries and do much more.
+
+https://github.com/MicrosoftDocs/Virtualization-Documentation/tree/live/windows-server-container-tools/CleanupContainerHostNetworking
+
+### Level 4 decontamination procedure
+
+All hope is lost; all we can do is hope that the contamination won't spread to other hosts.
+
+> Warning: this might BSOD.
+
+```
+// perform as single command because you will lose connectivity during netcfg -D
+netcfg -D; Restart-Computer -force
+```
+
+# HNS error specimen list
+
+The following chapter contains a list of HNS errors encountered throughout development, along with reproduction methods and natural habitat.
 
 ## 1. HNS Unspecified Error
 
+### Natural habitat:
+
+TODO
+
+### Reproduction:
+
+TODO
+
+## 2. HNS Invalid Parameter
+
+### Natural habitat:
+
+TODO
+
+### Reproduction:
+
+TODO
+
+## 3. HNS Element not found
+
+### Natural habitat:
+
 1. When attempting to create a transparent docker network
```

Comment:
I don't get it. This line was under `HNS Unspecified Error` before and it's under `HNS Element not found` now. Was it supposed to be like that?

Reply:
I think it belongs to this section, although we should mark it as hypothesis, because we're not sure if that's the cause of this error.

Reply:
It's HNS Element not found both in linked issue and on confluence on "Issues with Docker and HNS" page. I'm 90% sure that it's Element not found.